File: src/main/java/axirassa/services/email/EmailTemplate.java

package axirassa.services.email;

import lombok.Getter;
import axirassa.services.util.TemplateEnumeration;

public enum EmailTemplate implements TemplateEnumeration {

	USER_VERIFY_ACCOUNT("account/VerifyAccountEmail", EmailSenderAddress.ACCOUNT),
	USER_RESET_PASSWORD("account/ResetPasswordEmail", EmailSenderAddress.ACCOUNT),
	USER_CHANGE_PASSWORD("account/PasswordChangeEmail", EmailSenderAddress.ACCOUNT),
	USER_PRIMARY_EMAIL_CHANGED("account/PrimaryEmailChangedEmail", EmailSenderAddress.ACCOUNT),
	AGGREGATED_FEEDBACK("support/AggregatedFeedbackEmail", EmailSenderAddress.INTERNAL),
	USER_ADD_EMAIL("support/EmailAddedEmail", EmailSenderAddress.ACCOUNT);

	public static final String BASE_LOCATION = "/axirassa/webapp/emails/";

	@Getter
	private final String location;

	@Getter
	private final String fromAddress;


	EmailTemplate(String location, EmailSenderAddress fromAddress) {
		this.location = location;
		this.fromAddress = fromAddress.getAddress();
	}


	@Override
	public String getFullLocation() {
		return BASE_LOCATION + location;
	}
}

File: src/main/java/axirassa/services/email/EmailTemplate.scala
package axirassa.services.email

import axirassa.services.util.TemplateType
import axirassa.services.util.TemplateFactory

import scala.reflect.BeanProperty

object EmailTemplateType {
    val HTML = new EmailTemplateType("html")
    val TEXT = new EmailTemplateType("text")
    val SUBJECT = new EmailTemplateType("subject")
}
class EmailTemplateType(extension : String) extends TemplateType

object EmailTemplate {
    val USER_VERIFY_ACCOUNT = new EmailTemplate("account/VerifyAccountEmail", EmailSenderAddress.ACCOUNT)
    val USER_RESET_PASSWORD = new EmailTemplate("account/ResetPasswordEmail", EmailSenderAddress.ACCOUNT)
    val USER_CHANGE_PASSWORD = new EmailTemplate("account/PasswordChangeEmail", EmailSenderAddress.ACCOUNT)
    val USER_PRIMARY_EMAIL_CHANGED = new EmailTemplate("account/PrimaryEmailChangedEmail", EmailSenderAddress.ACCOUNT)
    val AGGREGATED_FEEDBACK = new EmailTemplate("support/AggregatedFeedbackEmail", EmailSenderAddress.INTERNAL)

    val BASE_LOCATION = "/axirassa/webapp/emails/"

}
class EmailTemplate(location : String, @BeanProperty fromAddress : String) extends {
    override def getFullLocation = EmailTemplate.BASE_LOCATION + location
}

class EmailTemplateFactory
    extends TemplateFactory[EmailTemplate, EmailTemplateType](EmailTemplate.BASE_LOCATION)

object EmailTemplateFactory {
    val instance = new EmailTemplateFactory

    def getTemplateInstance(template : EmailTemplate, templateType : EmailTemplateType) =
        instance.getTemplate(template, templateType)
}